user/stubs.c

```diff
@@ -0,0 +1,119 @@
+#include "userargs.h"
+#include <sys/syscall.h>
+
+sgxlkl_userargs_t* __sgxlkl_userargs;
+
+void sgxlkl_warn(const char* fmt, ...);
+
+int snprintf(char *str, size_t size, const char *format, ...);
+
+/*
+**==============================================================================
+**
+** syscall:
+**
+**==============================================================================
+*/
+
+long lkl_syscall(long no, long* params)
+{
+    long ret = __sgxlkl_userargs->ua_lkl_syscall(no, params);
+
+    return ret;
+}
+
+long __sgxlkl_log_syscall(
+    // sgxlkl_syscall_kind type,
+    uint32_t type,
+    long n,
+    long res,
+    int params_len,
+    ...)
+{
+    sgxlkl_warn("__sgxlkl_log_syscall() unimplemented in user space");
+    return 0;
+}
+
+/*
+**==============================================================================
+**
+** bypasses:
+**
+**==============================================================================
+*/
+
+void sgxlkl_warn(const char* msg, ...)
+{
+    /* ATTN: ignore variadic arguments */
+    return __sgxlkl_userargs->ua_sgxlkl_warn(msg);
+}
+
+void sgxlkl_error(const char* msg, ...)
+{
+    /* ATTN: ignore variadic arguments */
+    return __sgxlkl_userargs->ua_sgxlkl_error(msg);
+}
+
+void sgxlkl_fail(const char* msg, ...)
+{
+    /* ATTN: ignore variadic arguments */
+    return __sgxlkl_userargs->ua_sgxlkl_fail(msg);
+}
+
+bool sgxlkl_in_sw_debug_mode()
+{
+    return __sgxlkl_userargs->sw_debug_mode;
+}
+
+void* enclave_mmap(
+    void* addr,
+    size_t length,
+    int mmap_fixed,
+    int prot,
+    int zero_pages)
+{
+    return __sgxlkl_userargs->ua_enclave_mmap(addr, length, mmap_fixed,
+        prot, zero_pages);
```

Comment:
Userspace's libc shouldn't issue this call directly.  We probably need to still call `enclave_mmap` because the kernel libc tries to mmap things before LKL is running, but this one should just do an mmap syscall.
```suggestion
    return mmap(addr, length, prot, MAP_PRIVATE | MAP_ANONYMOUS |  (mmap_fixed ? MAP_FIXED : 0, -1, 0); 
```

Reply:
This is called in two places in malloc, neither sets `map_fixed`, so we don't actually need to handle this case.  It might be worth adding an assertion, something like:
```suggestion
    assert(prot == PROT_READ | PROT_WRITE);
    assert(!mmap_fixed);
    return mmap(addr, length, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0); 
```

This can then be removed once we are no longer linking a copy of libc into the kernel.